internal/extensions/extensions.go

```diff
@@ -0,0 +1,213 @@
+}
+
+// NewExtensionRegistry creates a new instance of the Extensions struct, which represents a registry of extensions.
+// We have the option to populate the internal extensions or not. Having s option becomes handy
+// when we want to create a custom registry of extensions (see `NewExtensionRegistryFromList`)
+func NewExtensionRegistry(populateInternal bool) (*Extensions, error) {
+	extensions := Extensions{}
+	var err error
+	if populateInternal {
+		extensions = make(Extensions, 0, len(*internalExtensions))
+		for _, internalExtension := range *internalExtensions {
+			err = validateInternalExtension(internalExtension)
+			if err != nil {
+				return nil, err
+			}
+
+			extensions = append(extensions, internalExtension)
+		}
+	}
+
+	return &extensions, nil
+}
+
+// NewExtensionRegistryFromList creates a new instance of the Extensions struct, from a list of extensions.
+// The passed extensions could be internal or external or just invalid.
+// The function is typically used to create a registry from an API response in order to compare
+// the source and the target systems' extensions for compatibility.
+// The returned registry is always sorted with internal extensions first.
+func NewExtensionRegistryFromList(extensions []string) (*Extensions, error) {
+	registry, err := NewExtensionRegistry(false)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, extension := range extensions {
+		err = validateExternalExtension(extension)
+		if err != nil {
+			// The extension could be an internal extension.
+			err = validateInternalExtension(extension)
+			if err != nil {
+				// The extension is invalid in both case.
+				return nil, fmt.Errorf("Extension %q is invalid", extension)
+			}
+
+			// The extension is an internal extension.
+			*registry = append(*registry, extension)
+		} else {
+			// The extension is an external extension.
+			*registry = append(*registry, extension)
+		}
+	}
+
+	return registry, nil
+}
+
+// Register registers new external extensions to the Extensions struct.
+func (e *Extensions) Register(newExtensions []string) error {
+	// Check for duplicates for internal and External extensions
+	for _, extension := range newExtensions {
+		if shared.ValueInSlice[string](extension, []string(*e)) {
+			return fmt.Errorf("Extension %q already registered", extension)
+		}
+
+		err := validateExternalExtension(extension)
+		if err != nil {
+			return err
+		}
+
+		*e = append(*e, extension)
+	}
+
+	return nil
+}
+
+// MarshalJSON implement the json.Marshaler interface to serialize the Extensions struct
+// for API responses.
+func (e *Extensions) MarshalJSON() ([]byte, error) {
+	return json.Marshal([]string(*e))
+}
+
+// UnmarshalJSON implement the json.Unmarshaler interface to deserialize the Extensions struct
+// from API requests.
+func (e *Extensions) UnmarshalJSON(data []byte) error {
+	var tmp []string
+	if err := json.Unmarshal(data, &tmp); err != nil {
+		return err
+	}
+
+	*e = Extensions(tmp)
+	return nil
+}
```

Comment:
Now that `Extensions` is a `[]string` I'm not sure we need the `json.Marshaler` `json.Unmarshaler` implementations as I believe the `json` library will understand the underlying type.

I could be wrong though

Reply:
fixed